### PR TITLE
2190: Setting maximum size limit for PR diff in generating Webrev

### PR DESCRIPTION
**Pull request by zhaosongzs**
https://github.com/openjdk/tsan/pull/17, there are more than 5000 commits in this pr and mlbridge bot was trying to generate the webrev. The generation of webrev was successful but the webrev file was too big and exceeded GitHub's file size limit.

So as ErikJ suggested, if the diff of the pr is too large, we shouldn't generate webrev for it.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-2190](https://bugs.openjdk.org/browse/SKARA-2190): Setting maximum size limit for PR diff in generating Webrev (**Bug** - P4)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara.git pull/1617/head:pull/1617` \
`$ git checkout pull/1617`

Update a local copy of the PR: \
`$ git checkout pull/1617` \
`$ git pull https://git.openjdk.org/skara.git pull/1617/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1617`

View PR using the GUI difftool: \
`$ git pr show -t 1617`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1617.diff">https://git.openjdk.org/skara/pull/1617.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/skara/pull/1617#issuecomment-1982142763)

### CODE REVIEW

File: webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java [hunk 1]
```diff
@@ -201,6 +201,16 @@ private boolean hasMergeCommits(Hash tailEnd, Hash head) throws IOException {
...
 
         private void generateJSON(Diff diff, Hash tailEnd, Hash head) throws IOException {
+            var totalChanges = diff.patches().stream()
+                    .filter(Patch::isTextual)
+                    .map(Patch::asTextualPatch)
+                    .flatMap(textualPatch -> textualPatch.hunks().stream())
+                    .mapToInt(Hunk::changes)
+                    .sum();
+            if (totalChanges > 100000) {
+                return;
```
Comment (zhaosongzs) [inline webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java:211]: 100000 is an arbitrary number. Please advice on this number.

File: bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java [hunk 2]
```diff
@@ -325,7 +325,9 @@ static String composeConversationFooter(PullRequest pr, URI issueProject, String
                 "Commit messages:\n" +
                 formatCommitMessagesBrief(commits, commitsLink).orElse("") + "\n\n" +
                 "Changes: " + pr.changeUrl() + "\n" +
-                (webrev.uri() == null ? "" : " Webrev: " + webrev.uri().toString() + "\n") +
+                (webrev.diffTooLarge() ?
+                        "  Webrev: Webrev is not available because diff is too large\n" :
```
Comment (erikj79) [inline bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java:329]: This string is repeated a lot, can we make it a constant, at least within this class?

File: webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java [hunk 3]
```diff
@@ -200,7 +200,20 @@ private boolean hasMergeCommits(Hash tailEnd, Hash head) throws IOException {
...
 
-        private void generateJSON(Diff diff, Hash tailEnd, Hash head) throws IOException {
+        private boolean diffTooLarge(Diff diff) {
+            var totalChanges = diff.patches().stream()
+                    .filter(Patch::isTextual)
+                    .map(Patch::asTextualPatch)
+                    .flatMap(textualPatch -> textualPatch.hunks().stream())
+                    .mapToInt(Hunk::changes)
+                    .sum();
+            return totalChanges > 300000;
```
Comment (erikj79) [inline webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java:210]: Please make this a constant.

File: bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java (hunk 2 above)
Comment (zhaosongzs) [reply bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java:329]: Sure, will fix it

File: webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java (hunk 3 above)
Comment (zhaosongzs) [reply webrev/src/main/java/org/openjdk/skara/webrev/Webrev.java:210]: Sure

**Review by erikj79 — APPROVED** [review]

**Comment by zhaosongzs** [comment]
Thanks for the review!
/integrate

**Comment by erikj79** [comment]
Test comment, please ignore.

**Pull request closed without merging**